syntax = "proto3";

package order_exchange;

service OrderExchangeSimulator {
  // Submit orders in batch
  rpc SubmitOrders(BatchOrderRequest) returns (BatchOrderResponse);
  
  // Cancel orders in batch
  rpc CancelOrders(BatchCancelRequest) returns (BatchCancelResponse);
}

message OrderRequest {
  string symbol = 1;
  enum Side {
    BUY = 0;
    SELL = 1;
  }
  Side side = 2;
  double quantity = 3;
  double price = 4;
  enum Type {
    MARKET = 0;
    LIMIT = 1;
  }
  Type type = 5;
  string request_id = 6;
}

message BatchOrderRequest {
  repeated OrderRequest orders = 1;
}

message OrderResponse {
  bool success = 1;
  string order_id = 2;
  string error_message = 3;
}

message BatchOrderResponse {
  bool success = 1;
  repeated OrderResponse results = 2;
  string error_message = 3;
}

message BatchCancelRequest {
  repeated string order_ids = 1;
}

message CancelResult {
  string order_id = 1;
  bool success = 2;
  string error_message = 3;
}

message BatchCancelResponse {
  bool success = 1;
  repeated CancelResult results = 2;
  string error_message = 3;
}